Add support for sparse_reduce_max and sparse_reduce_max_sparse

yongtang opened the pull request:
This commit tries to address the issue raised in #10002 to have
the support for sparse_reduce_max and sparse_reduce_max_sparse.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add support for sparse_reduce_max and sparse_reduce_max_sparse

This commit tries to address the issue raised in 10002 to have
the support for sparse_reduce_max and sparse_reduce_max_sparse.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

concretevitamin commented:
@yongtang I was imagining the code for `sparse_reduce_sum_op.cc` and `sparse_reduce_max_op.cc` can be mostly shared. A remotely related example is the `ScatterUpdateOp`, which has a template param on `UpdateOp` (`ADD`, etc.).

yongtang committed:
Move sparse_reduce_sum_op.cc to sparse_reduce_op.cc

This commit moves sparse_reduce_sum_op.cc to sparse_reduce_op.cc
so that both `sparse_reduce_sum` and `sparse_reduce_max` share
the same code base.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @concretevitamin for the review. The PR has been updated so that `sparse_reduce_sum` and `sparse_reduce_max` share the same code base. Please take a look.

concretevitamin commented on tensorflow/core/kernels/sparse_reduce_op.cc:
```diff
@@ -195,7 +209,7 @@ class SparseReduceSumOp : public OpKernel {
     sp.Reorder<T>(reduction.reorder_dims);
     for (const auto &g : sp.group(reduction.group_by_dims)) {
-      group_sum.device(ctx->eigen_cpu_device()) = g.template values<T>().sum();
+      Op::template Run<T>(ctx, group_sum, g.template values<T>());
```
Rename `group_sum`?  Something like `reduced_val` or something better.

yongtang replied:
Thanks @concretevitamin. The PR has been updated.

concretevitamin commented on tensorflow/core/kernels/sparse_reduce_op.cc:
```diff
@@ -265,7 +286,7 @@ class SparseReduceSumSparseOp : public OpKernel {
     int64 i = 0;
     for (const auto &g : sp.group(reduction.group_by_dims)) {
-      group_sum.device(ctx->eigen_cpu_device()) = g.template values<T>().sum();
+      Op::template Run<T>(ctx, group_sum, g.template values<T>());
```
ditto

yongtang replied:
Done.

concretevitamin commented on tensorflow/core/kernels/sparse_reduce_max_op_test.cc:
```diff
@@ -0,0 +1,125 @@
+/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.
```
I'd prefer if you could leave out this file.  The reason we originally had a C++ test file for `SparseReduceSum` was to ease debugging.  Since now most of the kernels share the same logic, this is no longer needed.  The functional correctness of the new op can be covered by the more lightweight Python unit tests.

yongtang replied:
Thanks. Done.

concretevitamin commented on tensorflow/python/kernel_tests/sparse_ops_test.py:
```diff
@@ -510,10 +510,86 @@ def testNoEmptyRows(self):
       self.assertAllEqual(empty_row_indicator_out, np.zeros(2).astype(np.bool))
 
 
+class SparseReduceMaxTest(test_util.TensorFlowTestCase):
```
Can you refactor the existing `SparseReduceSumTest` to plumb through an extra bool argument, such as `do_sum == True / False`, so that we can share most of the existing test logic?

yongtang replied:
Thanks. Done.

concretevitamin reviewed (changes requested):
This looks great!  Left a few comments.

yongtang commented:
@concretevitamin Thanks for the review. The PR has been updated. Please take.a look.

concretevitamin commented on tensorflow/core/kernels/sparse_reduce_op.cc:
```diff
@@ -129,10 +129,30 @@ Status ValidateInputs(const Tensor *shape_t, const Tensor *reduction_axes_t) {
+  static void Run(OpKernelContext *ctx, typename TTypes<T>::Scalar &s, const typename TTypes<T>::UnalignedVec &v) {
+      s.device(ctx->eigen_cpu_device()) = v.sum();
+  }
+  static const char * Name() {
```
Let's do 
```cpp
StringPiece Name() {
    return "sum";
}
```
which is more succinct (and potentially safer).

concretevitamin commented on tensorflow/core/kernels/sparse_reduce_op.cc:
```diff
@@ -129,10 +129,30 @@ Status ValidateInputs(const Tensor *shape_t, const Tensor *reduction_axes_t) {
+  static void Run(OpKernelContext *ctx, typename TTypes<T>::Scalar &s, const typename TTypes<T>::UnalignedVec &v) {
+      s.device(ctx->eigen_cpu_device()) = v.maximum();
+  }
+  static const char * Name() {
```
ditto

concretevitamin reviewed (approved):
Great!  Two final nits.  Please address but I'm approving first ;)

yongtang commented:
Thanks @concretevitamin. The PR has been updated.

concretevitamin commented:
Jenkins, test this please.

@andrewharp feel free to merge.

jhseu commented:
This change will require API review.

josh11b commented:
Fine for API review.

martinwicke commented:
Can you rebase and fix the conflicts? I cannot fix things in this PR.

yongtang commented:
Thanks @martinwicke. The PR has been rebased. Please take a look

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
Can you please run the API goldens updater as described in the error message of the failing Linux-CPU test?

yongtang committed:
Update API goldens for API compatibility

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@martinwicke The API goldens has been updated with the PR. Please take a look and let me know if there are any issues.

martinwicke commented:
Jenkins, test this please.
